### ISSUE TITLE
Unused generic parameter

### ISSUE THREAD

**Issue by woshilapin** [rust-lang/rust-clippy#4610]
I ended up with a weird case of unused generic parameter that was not detected by the compilation (like it usually is) and that clippy didn't saw either. I'm not sure why but I've been able to craft a Minimal Example (or at least, I was not able to remove any part without the compiler to yell at me).

```rust
trait Foo {                   
    fn f();                
}                             
struct Bar;                   
                              
impl<'a> Foo for &'a Bar {    
    fn f() {               
        println!("foobar");   
    }                         
}                             
                              
fn my_func<H>()               
where                         
    for<'a> &'a H: Foo,       
{                             
    println!("Hello, world!");
}                             
                              
fn main() {                   
    my_func();                
}                             
```
In the above example, the function `my_func` has a generic parameter `H` that is never used and neither the compiler nor clippy is mentioning it.

Looking for other related issues, I've found #877 but the bug has been fixed. And like in #877, it might be a use case for `rustc` directly, I'm not sure.

**Comment by sinkuu** [comment]
AFAIK rustc never warns against unused type parameters on functions.
```rust
fn foo<T>() {} // No warning against `T`
```

**Comment by woshilapin** [comment]
Indeed, thank you for the precision.  Then it might be a good case for clippy?

**Comment by Areredify** [comment]
I would like to tackle this, no promises though
